src/trajectory/util.cpp

```diff
@@ -0,0 +1,314 @@
+      if (segmentDuration > 0.0)
+      {
+        CubicSplineProblem problem(Vector2d{0., segmentDuration}, 4, dimension);
+        problem.addConstantConstraint(0, 0, zeroPos);
+        problem.addConstantConstraint(0, 1, segStartVel);
+        problem.addConstantConstraint(1, 0, segEndPos - segStartPos);
+        problem.addConstantConstraint(1, 1, segEndVel);
+        const auto solution = problem.fit();
+        const auto coefficients = solution.getCoefficients().front();
+        outputTrajectory->addSegment(
+            coefficients, segmentDuration, segmentStartState);
+      }
+      break;
+    }
+
+    currSegmentIdx++;
+    currSegmentStartTime = currSegmentEndTime;
+  }
+
+  for (std::size_t i = currSegmentIdx + 1u; i < traj.getNumSegments(); i++)
+  {
+    outputTrajectory->addSegment(
+        traj.getSegmentCoefficients(i),
+        traj.getSegmentDuration(i),
+        traj.getSegmentState(i));
+  }
+
+  return outputTrajectory;
+}
+
+//==============================================================================
+UniqueInterpolatedPtr convertToInterpolated(
+    const Spline& traj, statespace::ConstInterpolatorPtr interpolator)
+{
+  auto stateSpace = traj.getStateSpace();
+  auto outputTrajectory
+      = make_unique<Interpolated>(stateSpace, std::move(interpolator));
+
+  auto state = stateSpace->createState();
+  for (std::size_t i = 0; i < traj.getNumWaypoints(); i++)
+  {
+    traj.getWaypoint(i, state);
+    const double t = traj.getWaypointTime(i);
+    outputTrajectory->addWaypoint(t, state);
+  }
+
+  return outputTrajectory;
+}
+
+//==============================================================================
+UniqueInterpolatedPtr concatenate(
+    const Interpolated& traj1, const Interpolated& traj2)
+{
+  if (traj1.getStateSpace()->getDimension()
+      != traj2.getStateSpace()->getDimension())
+    throw std::runtime_error("Dimension mismatch");
+
+  auto stateSpace = traj1.getStateSpace();
+
+  auto outputTrajectory
+      = make_unique<Interpolated>(stateSpace, traj1.getInterpolator());
+
+  for (std::size_t i = 0; i < traj1.getNumWaypoints() - 1; i++)
+  {
+    outputTrajectory->addWaypoint(
+        traj1.getWaypointTime(i), traj1.getWaypoint(i));
+  }
+  outputTrajectory->addWaypoint(traj1.getEndTime(), traj2.getWaypoint(0));
+  double endTimeOfTraj1 = traj1.getEndTime();
+  for (std::size_t i = 1; i < traj2.getNumWaypoints(); i++)
+  {
+    outputTrajectory->addWaypoint(
+        endTimeOfTraj1 + traj2.getWaypointTime(i), traj2.getWaypoint(i));
+  }
+  return outputTrajectory;
+}
+
+//==============================================================================
+UniqueSplinePtr concatenate(const Spline& traj1, const Spline& traj2)
+{
+  if (traj1.getStateSpace() != traj2.getStateSpace())
+    throw std::runtime_error("State space mismatch");
+  auto stateSpace = traj1.getStateSpace();
+  statespace::ConstInterpolatorPtr interpolator
+      = std::make_shared<statespace::GeodesicInterpolator>(stateSpace);
```

Comment:
Same as in `concatenate(Interpolated, Interpolated)`: is there a reason to define `stateSpace` above rather than just calling it here?